src/GoTrueClient.ts

```diff
@@ -1056,148 +1062,121 @@ export default class GoTrueClient {
-    const { data, error } = await this.getUser()
-    if (error) throw error
-    const user: User = data.user
-    try {
-      return await _request(this.fetch, 'POST', `${this.url}/user/${user.id}/factor`, {
-        body: {
-          friendly_name: params.friendlyName,
-          factor_type: factorType,
-          issuer: params.issuer,
-        },
-        headers: this.headers,
-      })
-    } catch (error) {
-      if (isAuthError(error)) {
-        return { data: null, error }
-      }
-      throw error
+  private async _enroll(params: MFAEnrollParams): Promise<AuthMFAEnrollResponse> {
+    const { data: sessionData, error: sessionError } = await this.getSession()
+    if (sessionError) {
+      return { data: null, error: sessionError }
     }
+
+    return await _request(this.fetch, 'POST', `${this.url}/factors`, {
+      body: {
+        friendly_name: params.friendlyName,
+        factor_type: params.factorType,
+        issuer: params.issuer,
+      },
+      headers: this.headers,
+      jwt: sessionData?.session?.access_token,
+    })
   }
 
   /**
    * Validates a device as part of the enrollment step.
    * @param factorID System assigned identifier for authenticator device as returned by enroll
    * @param code Code Generated by an authenticator device
    */
-  private async _verify(params: MFAVerifyParams): Promise<AuthResponse> {
-    const { data, error } = await this.getUser()
-    if (error) throw error
-    const user: User = data.user
-    try {
-      const { data, error } = await _request(
-        this.fetch,
-        'POST',
-        `${this.url}/user/${user.id}/factor/${params.factorId}/verify`,
-        {
-          body: { code: params.code, challenge_id: params.challengeId },
-          xform: _sessionResponse,
-          headers: this.headers,
-        }
-      )
-      if (error || !data) return { data: { user: null, session: null }, error }
-      if (data.session) {
-        await this._saveSession(data.session)
-        this._notifyAllSubscribers('MFA_CHALLENGE_VERIFIED', data.session)
-      }
-      return { data, error }
-    } catch (error) {
-      if (isAuthError(error)) {
-        return { data: { user: null, session: null }, error }
-      }
+  private async _verify(params: MFAVerifyParams): Promise<AuthMFAVerifyResponse> {
+    const { data: sessionData, error: sessionError } = await this.getSession()
+    if (sessionError) {
+      return { data: null, error: sessionError }
+    }
 
-      throw error
+    const { data, error } = await _request(
+      this.fetch,
+      'POST',
+      `${this.url}/factors/${params.factorId}/verify`,
+      {
+        body: { code: params.code, challenge_id: params.challengeId },
+        headers: this.headers,
+        jwt: sessionData?.session?.access_token,
+      }
+    )
+    if (error) {
+      return { data: null, error }
     }
+
+    await this._saveSession({
+      expires_at: Math.round(Date.now() / 1000) + data.expires_in,
+      ...data,
```

Comment:
`_saveSession` would not save it unless it has an `expires_at`, this is the best I could come up with. Please double check @kangmingtay @J0.